tee-worker/identity/litentry/core/direct-call-request/receiver/src/types.rs

```diff
@@ -0,0 +1,128 @@
+	pub ocall_api: Arc<OCallApi>,
+	pub data_provider_config: Arc<DataProviderConfig>,
+	pub extrinsic_factory: Arc<ExtrinsicFactory>,
+	pub node_metadata_repo: Arc<NodeMetadataRepo>,
+}
+
+impl<
+		ShieldingKeyRepository,
+		AuthorApi,
+		StfEnclaveSigning,
+		OCallApi,
+		ExtrinsicFactory,
+		NodeMetadataRepo,
+	>
+	DirectCallRequestContext<
+		ShieldingKeyRepository,
+		AuthorApi,
+		StfEnclaveSigning,
+		OCallApi,
+		ExtrinsicFactory,
+		NodeMetadataRepo,
+	> where
+	ShieldingKeyRepository: AccessKey + Send + Sync + 'static,
+	<ShieldingKeyRepository as AccessKey>::KeyType: ShieldingCryptoEncrypt + ShieldingCryptoDecrypt,
+	AuthorApi: AuthorApiTrait<Hash, Hash, TrustedCallSigned, Getter> + Send + Sync + 'static,
+	StfEnclaveSigning: StfEnclaveSigningTrait<TrustedCallSigned> + Send + Sync + 'static,
+	OCallApi:
+		EnclaveOnChainOCallApi + EnclaveMetricsOCallApi + EnclaveAttestationOCallApi + 'static,
+	ExtrinsicFactory: CreateExtrinsics + Send + Sync + 'static,
+	NodeMetadataRepo: AccessNodeMetadata + Send + Sync + 'static,
+	NodeMetadataRepo::MetadataType: NodeMetadataTrait,
+{
+	#[allow(clippy::too_many_arguments)]
+	pub fn new(
+		shielding_key: Arc<ShieldingKeyRepository>,
+		author_api: Arc<AuthorApi>,
+		enclave_signer: Arc<StfEnclaveSigning>,
+		enclave_account: Arc<Ed25519Pair>,
+		ocall_api: Arc<OCallApi>,
+		data_provider_config: Arc<DataProviderConfig>,
+		extrinsic_factory: Arc<ExtrinsicFactory>,
+		node_metadata_repo: Arc<NodeMetadataRepo>,
+	) -> Self {
+		Self {
+			shielding_key,
+			author_api,
+			enclave_signer,
+			enclave_account,
+			ocall_api,
+			data_provider_config,
+			extrinsic_factory,
+			node_metadata_repo,
+		}
+	}
+}
+
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq)]
+pub enum DirectCallErrorDetail {
+	UnexpectedCall(String),
+	ShieldingKeyRetrievalFailed(String), // Stringified itp_sgx_crypto::Error
+	RequestPayloadDecodingFailed,
+	ParentchainDataRetrievalFailed(String), // Stringified itp_stf_state_handler::Error
+	InvalidSignerAccount,
+	UnauthorizedSigner,
+	MissingAesKey,
+	MrEnclaveRetrievalFailed,
+	EnclaveSignerRetrievalFailed,
+	SignatureVerificationFailed,
+	ConnectionHashNotFound(String),
+	MetadataRetrievalFailed(String), // Stringified itp_node_api_metadata_provider::Error
+	InvalidMetadata(String),         // Stringified itp_node_api_metadata::Error
+	TrustedCallSendingFailed(String), // Stringified mpsc::SendError<(H256, TrustedCall)>
+	CallSendingFailed(String),
+	ExtrinsicConstructionFailed(String), // Stringified itp_extrinsics_factory::Error
+	ExtrinsicSendingFailed(String),      // Stringified sgx_status_t
+}
+
+// split this type and move it to the right place once we have the new architecture
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq)]
+pub struct DirectCallResult {
+	pub account_store_hash: H256,
+	pub mutated_account_store: AesOutput,
+}
+
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq)]
+pub struct DirectCallRequestResult {
```

Comment:
Maybe we don't need this extra wrapper - can we use `Result< , >` directly?

Reply:
I think it would be better, I was using the same structure we have for `vc` (`RequestVcResultOrError`) but we probably don't need this.